xgboost_ray/main.py

```diff
@@ -800,7 +824,37 @@ def train(params: Dict,
             "`dtrain = RayDMatrix(data=data, label=label)`.".format(
                 type(dtrain)))
 
-    _try_add_tune_callback(kwargs)
+    if not ray.is_initialized():
+        ray.init()
+
+    gpus_per_actor = ray_params.gpus_per_actor
+    cpus_per_actor = ray_params.cpus_per_actor
+
+    # Automatically set gpus_per_actor if left at the default value
+    if gpus_per_actor == -1:
+        gpus_per_actor = 0
+        if "tree_method" in params and params["tree_method"].startswith("gpu"):
+            gpus_per_actor = 1
+
+    # Automatically set cpus_per_actor if left at the default value
+    # Will be set to the number of cluster CPUs divided by the number of
+    # actors, bounded by the maximum number of CPUs across actors nodes.
+    if cpus_per_actor <= 0:
+        cluster_cpus = _ray_get_cluster_cpus() or 1
+        cpus_per_actor = min(
+            int(_get_min_node_cpus() or 1),
```

Comment:
Btw I think this should be `_get_min_node_cpus`, not `_get_max_node_cpus`. If we have a cluster with 3 nodes: 1 with 1 CPU, and 2 with 3 CPUS each, and we have num_actors=4, how many cpus should we have per actor?

Previously this would give 2 cpus per actor, but that's currently not possible with this setup. We should be bounded by the _minimum_ number of CPUs not the maximum, if I'm understanding this correctly.

Reply:
Hmm, ideally you end up with 3 actors, but I think you'll want to do 3 CPUs each per actor (resulting in 2 created actors). 

The reason is because xgboost is largely memory bound, and doubling 2 actors onto 1 machine will be less desirable (increases the likelihood of OOM).

Reply:
Right, but we currently don't automatically configure the number of actors, only the number of _cpus/gpus_ per actor. So if the the user passes in `num_actors=4`, we will always create 4 actors even if it's undesirable.

Also, the example cluster setup I gave was incorrect, it should be 4 nodes in total, 2 with 1 CPU each, 2 with 3 CPUs each. 

In master, autodetecting CPUs would set `num_cpus_per_actor` to 2 with this setup:
`_get_max_node_cpus()` would give 3,
`int(cluster_cpus // num_actors)` == `int(8 // 4)` which would give 2,
and we would take the min of these to get 2 cpus per actor.
But having 4 actors with 2 CPUs each is not able to be scheduled with this cluster setup.

Instead we want to have `get_min_node_cpus` which would return 1. So we would have 4 actors with 1 CPU each which can be scheduled.

Reply:
I see; that makes sense.